### PR TITLE
Add Support for Cloud Native Buildpacks API

### PR DESCRIPTION
**Pull request by jkutner**
* add `bin/build`
* move common functions into `lib/`
* add `buildpack.toml` elements for v3/cnb
* refactor test
* removed some lingering JDK 6 references

### CODE REVIEW

File: bin/compile [hunk 1]
```diff
@@ -11,4 +11,4 @@ BIN_DIR=$(cd $(dirname $0); pwd) # absolute path
 . $BIN_DIR/util
 
 # Install JDK
-install_java_with_overlay ${BUILD_DIR}
+install_java_with_overlay "${BUILD_DIR}" "${CACHE_DIR}"
```
Comment (jkutner) [inline bin/compile]: this is unrelated to CNB and just something the v2 buildpack should have been doing

**Comment by schneems** [comment]
Removing the label until you think this is ready for review

**Comment by jkutner** [comment]
@schneems it is ready for review

File: buildpack.toml [hunk 2]
```diff
@@ -14,3 +19,6 @@ name = "JVM"
     "Gemfile",
     "Gemfile.lock"
   ]
+
+[[stacks]]
+id = "heroku-18"
```
Comment (CaseyFaist) [inline buildpack.toml]: Should this have a newline at EOF?

File: buildpack.toml [hunk 3]
```diff
@@ -1,6 +1,11 @@
+api = "0.2"
+
 [buildpack]
+id = "heroku/jvm"
+version = "0.1"
 name = "JVM"
 
+
```
Comment (CaseyFaist) [inline buildpack.toml]: Extra space

File: lib/jvm.sh [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+  local file=${1:?}
+  local key=${2:?}
+
+  # escape for regex
+  local escaped_key=$(echo $key | sed "s/\./\\\./g")
+
+  [ -f $file ] && \
+  grep -E ^$escaped_key[[:space:]=]+ $file | \
+  sed -E -e "s/$escaped_key([\ \t]*=[\ \t]*|[\ \t]+)([_A-Za-z0-9\.-]*).*/\2/g"
+}
```
Comment (CaseyFaist) [inline lib/jvm.sh]: Newline at end of file

File: lib/v3/buildpack.sh [hunk 5]
```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env bash
+
+bp_layer_rebuild?() {
```
Comment (CaseyFaist) [inline lib/v3/buildpack.sh]: This looks like it will always return 0 and thus won't run - is that intended?

File: lib/v3/buildpack.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  local metadata="${3:?}"
+  
+  local layer_dir="${layers_dir}/${name}"
+  local layer_metadata="${layer_dir}.toml"
+
+  mkdir -p "${layer_dir}"
+  echo "${metadata}" > "${layer_metadata}"
+
+  echo "${layer_dir}"
+}
```
Comment (CaseyFaist) [inline lib/v3/buildpack.sh]: newline

File: test/v3 [hunk 7]
```diff
@@ -0,0 +1,60 @@
+#!/usr/bin/env bash
+
+export BUILDPACK_HOME="${BUILDPACK_HOME:-$(pwd)}"
+export CNB_STACK_ID="heroku"
```
Comment (hone) [inline test/v3]: Do you basically ignore this value? Should this be set by the `Makefile`?

File: Makefile [hunk 8]
```diff
@@ -0,0 +1,30 @@
+unit:
+	@echo "Running unit tests in docker (heroku-18)..."
+	@docker run -v $(shell pwd):/buildpack:ro --rm -it -e "STACK=heroku-18" -e "BUILDPACK_HOME=/buildpack" heroku/heroku:18 bash -c 'cp -r /buildpack /buildpack_test; cd /buildpack_test/; test/unit;'
+	@echo ""
+
+v3:
+	@echo "Running v3 integration tests in docker (heroku-18)..."
+	@docker run -v $(shell pwd):/buildpack:ro --rm -it -e "STACK=heroku-18" -e "BUILDPACK_HOME=/buildpack" heroku/heroku:18 bash -c 'cp -r /buildpack /buildpack_test; cd /buildpack_test/; test/v3;'
```
Comment (hone) [inline Makefile]: Should we be using `CNB_STACK_ID` instead of `STACK` for v3?

File: buildpack.toml [hunk 9]
```diff
@@ -1,4 +1,8 @@
+api = "0.2"
+
 [buildpack]
+id = "heroku/jvm"
+version = "0.1"
 name = "JVM"
 
   [publish.Ignore]
```
Comment (hone) [inline buildpack.toml]: in the long term, I wonder if having these extra keys that don't comply with `buildpack.toml` in v3 matter.

File: lib/v3/buildpack.sh [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+    curl -o "$yjBin" -Ls "$yjUrl" && chmod +x "$yjBin"
+    local actualSha="$(shasum -a 256 $yjBin | awk '{ print $1 }')"
+    if [ "$actualSha" != "$yjSha" ]; then
+      echo "Invalid yj sha: $actualSha"
+      exit 1
+    fi
+  fi
+
+  echo "cache = true" > "${layer_dir}.toml"
+  echo "build = true" >> "${layer_dir}.toml"
```
Comment (hone) [inline lib/v3/buildpack.sh]: do the other Java buildpacks expect this buildpack to setup the utils/tools? I wonder if this should just be its own buildpack.

File: bin/build [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+  install_jdk "${jdkVersion}" "${jdkUrl}" "${jdkLayer}"
+  install_certs "${jdkLayer}"
+  install_profile "${bpDir}" "${jdkLayer}/profile.d"
+  install_jdk_overlay "${jdkVersion}" "${jdkLayer}" "${appDir}"
+
+  info "Installing JRE ${jdkVersion}"
+  jreLayerToml="$(bp_layer_metadata_create "true" "false" "false" "${jdkMetadata}")"
+  jreLayer="$(bp_layer_init "${layersDir}" "jre" "${jreLayerToml}")"
+  install_jre "${jdkLayer}" "${jreLayer}"
```
Comment (hone) [inline bin/build]: should this JRE layer logic be treated and checked separately?

File: bin/build [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+  jdkLayerToml="$(bp_layer_metadata_create "false" "true" "true" "${jdkMetadata}")"
+  jdkLayer="$(bp_layer_init "${layersDir}" "jdk" "${jdkLayerToml}")"
+
+  install_jdk "${jdkVersion}" "${jdkUrl}" "${jdkLayer}"
+  install_certs "${jdkLayer}"
+  install_profile "${bpDir}" "${jdkLayer}/profile.d"
+  install_jdk_overlay "${jdkVersion}" "${jdkLayer}" "${appDir}"
+
+  info "Installing JRE ${jdkVersion}"
+  jreLayerToml="$(bp_layer_metadata_create "true" "false" "false" "${jdkMetadata}")"
```
Comment (hone) [inline bin/build]: Why would we not want to cache the JRE?

File: Makefile (hunk 8 above)
Comment (jkutner) [reply Makefile]: probably. at least `CNB_STACK_ID` maybe both

File: bin/build (hunk 11 above)
Comment (jkutner) [reply bin/build]: I don't think it's necessary because the JRE is contained within the JDK (so it's essentially extracted from that layer). They are very much entangled.

File: buildpack.toml (hunk 9 above)
Comment (jkutner) [reply buildpack.toml]: we could have buildpack registry look for them in the `[metadata]`

File: lib/v3/buildpack.sh (hunk 10 above)
Comment (jkutner) [reply lib/v3/buildpack.sh]: yea, i considered a `buildpack-utils` buildpack, but I think we can decide on that later (like we did with buildpack-stdlib). In any case, I'm leaning towards vendoring

File: test/v3 (hunk 7 above)
Comment (jkutner) [reply test/v3]: it should, and I should honor it instead of STACK in the v3 stuff

File: Makefile (hunk 8 above)
Comment (jkutner) [reply Makefile]: hmm, setting both when running v3 messes up the `bin/detect`, which checks for `CNB_STACK_ID`

Comment (jkutner) [reply Makefile]: setting CNB_STACK_ID for v3 only and STACK for v2 now

File: lib/v3/buildpack.sh (hunk 10 above)
Comment (hone) [reply lib/v3/buildpack.sh]: does that mean the other buildpacks in the chain fail if `build = false`?

File: bin/build (hunk 11 above)
Comment (hone) [reply bin/build]: Should the JRE layer contain the ETAG from the JRE tarball?

Comment (jkutner) [reply bin/build]: there is no JRE tarball. it's a subdir of the exploded JDK
